### PR TITLE
[SPARK-24195][Core] Ignore the files with "local" scheme in SparkContext.addFile

### PR DESCRIPTION
**Pull request by xuanyuanking**
## What changes were proposed in this pull request?

In Spark "local" scheme means resources are already on the driver/executor nodes, this pr ignore the files with "local" scheme in `SparkContext.addFile` for fixing potential bug.

## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by xuanyuanking** [comment]
cc @felixcheung. Please take a look about this when you have time. Thanks.

**Comment by gatorsmile** [comment]
cc @jiangxb1987 @jerryshao

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -1517,9 +1517,12 @@ class SparkContext(config: SparkConf) extends Logging {
    * only supported for Hadoop-supported filesystems.
    */
   def addFile(path: String, recursive: Boolean): Unit = {
-    val uri = new Path(path).toUri
+    var uri = new Path(path).toUri
     val schemeCorrectedPath = uri.getScheme match {
-      case null | "local" => new File(path).getCanonicalFile.toURI.toString
+      case null | "local" =>
+        // SPARK-24195: Local is not a valid scheme for FileSystem, we should only keep path here.
+        uri = new Path(uri.getPath).toUri
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Why is this needed? Can't we just do `new File(uri.getPath).getCanonicalFile.toURI.toString` without this line?

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 2]
```diff
@@ -116,49 +116,52 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext with Eventu
...
     val absolutePath1 = file1.getAbsolutePath
 
+    // file and absolute path for relative path
     val file2 = File.createTempFile("someprefix2", "somesuffix2", dir)
     val relativePath = file2.getParent + "/../" + file2.getParentFile.getName + "/" + file2.getName
     val absolutePath2 = file2.getAbsolutePath
 
+    // file and absolute path for path with local scheme
+    val file3 = File.createTempFile("someprefix3", "somesuffix3", dir)
+    val localPath = "local://" + file3.getParent + "/../" + file3.getParentFile.getName +
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: Let's use string interpolation.

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 3]
```diff
@@ -116,49 +116,52 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext with Eventu
...
-        if (absolutePath2 == gotten2.getAbsolutePath) {
-          throw new SparkException("file should have been copied : " + absolutePath2)
-        }
+      }
+
+      sc = new SparkContext(new SparkConf().setAppName("test").setMaster("local"))
+      sc.addFile(file1.getAbsolutePath)
+      sc.addFile(relativePath)
+      sc.addFile(localPath)
+      sc.parallelize(Array(1), 1).map(x => {
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: nit:

```
map { x =>
  ...
}
```

**Comment by HyukjinKwon** [comment]
Seems fine to me.

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 1 above)
Comment (jerryshao) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: Yes, same question. The above line seems not useful.

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 4]
```diff
@@ -116,49 +116,52 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext with Eventu
...
-              absolutePath2)
-        }
 
-        if (absolutePath1 == gotten1.getAbsolutePath) {
+        if (absolutePath == gotten.getAbsolutePath) {
           throw new SparkException("file should have been copied :" + absolutePath1)
         }
-        if (absolutePath2 == gotten2.getAbsolutePath) {
-          throw new SparkException("file should have been copied : " + absolutePath2)
-        }
```
Comment (felixcheung) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: can we not change the existing test?

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 1 above)
Comment (felixcheung) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: it changes `uri` - which is reference again below.

**Review by felixcheung — COMMENTED** [review]
can you give some examples of what the output for local:/ looks like in the change of addFile()?

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 1 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: Yes, just as @felixcheung said, this because we will use uri in https://github.com/apache/spark/pull/21533/files/f922fd8c995164cada4a8b72e92c369a827def16#diff-364713d7776956cb8b0a771e9b62f82dR1557, if the uri with local scheme, we'll get an exception cause local is not a valid scheme for FileSystem.

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala (hunk 2 above)
Comment (xuanyuanking) [reply core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: Got it, thanks.

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala (hunk 3 above)
Comment (xuanyuanking) [reply core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: Got it, fix in next commit.

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala (hunk 4 above)
Comment (xuanyuanking) [reply core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: Actually I keep all existing test and just do clean work for reducing common code line by adding a function checkGottenFile in https://github.com/apache/spark/pull/21533/files/f922fd8c995164cada4a8b72e92c369a827def16#diff-8d5858d578a2dda1a2edb0d8cefa4f24R139. If you think it's unnecessary, I just change it back.

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 1 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: I mean we `getPath` doesn't include scheme:

```
scala> new Path("local://a/b/c")
res0: org.apache.hadoop.fs.Path = local://a/b/c

scala> new Path("local://a/b/c").toUri
res1: java.net.URI = local://a/b/c

scala> new Path("local://a/b/c").toUri.getScheme
res2: String = local

scala> new Path("local://a/b/c").toUri.getPath
res3: String = /b/c
```

why should we do this again?

```
scala> new Path(new Path("local://a/b/c").toUri.getPath).toUri.getPath
res4: String = /b/c
```

Comment (jiangxb1987) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: Yea we can simplify this.

**Comment by jiangxb1987** [comment]
LGTM except for the comment from @HyukjinKwon

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 1 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: @HyukjinKwon @jiangxb1987 
Thanks for your explain, I think I know what's your meaning about `we getPath doesn't include scheme`. Actually the purpose of this code `uri = new Path(uri.getPath).toUri`, is to reassign the var in +1520, we don't want the uri including local scheme.
```
Can't we just do new File(uri.getPath).getCanonicalFile.toURI.toString without this line?
```
We can't because like I explained above, if we didn't do `uri = new Path(uri.getPath).toUri`, will get a exception like below:
```
No FileSystem for scheme: local
java.io.IOException: No FileSystem for scheme: local
	at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:2586)
	at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2593)
	at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:91)
	at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2632)
	at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2614)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:370)
	at org.apache.spark.util.Utils$.getHadoopFileSystem(Utils.scala:1830)
	at org.apache.spark.util.Utils$.doFetchFile(Utils.scala:690)
	at org.apache.spark.util.Utils$.fetchFile(Utils.scala:486)
	at org.apache.spark.SparkContext.addFile(SparkContext.scala:1557)
```

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: I mean, at least we can do:

```
val a = new File(uri.getPath).getCanonicalFile.toURI.toString
uri = new Path(uri.getPath).toUri
a
```

`new Path(uri.getPath).toUri` for trimming the scheme looks not quite clean though. It's a-okay at least to me.

**Comment by jerryshao** [comment]
Just take another look on this issue. I think the fix is just to make it work, but not make it work correctly.

The fix here and the original code actually treats scheme "local" to "file", actually they're different in Spark.

In Spark "local" scheme means resources are already on the driver/executor nodes, which means Spark doesn't need to ship resources from driver to executors via fileserver. But here it treats as "file" which will be shipped via fileserver to executors. This is semantically not correct.

I think for "local" scheme, the fix should:

1. Make it accessible both from driver and executors via `SparkFiles#get`. By copying resource to the folders.
2. It should not be added into fileServer.

**Comment by HyukjinKwon** [comment]
Exactly I agree with ^. It should be the best to implement that logic. I thought it seems we have never implemented "local" logic as described so far. So, I thought it's kind of okay.

**Comment by jerryshao** [comment]
"local" scheme was supported long ago for users who already deploy jars on every node. HDI heavily uses this feature.

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 1 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: Ah, I see, thanks. I'll do this in the next commit. Thanks for your patient explain.

**Comment by xuanyuanking** [comment]
@jerryshao Great thanks for your review and detailed explain, based on your guidance, I found the behavior about the file in `local` scheme added in fileServer was introduced by this PR: https://github.com/apache/spark/commit/c4b1108c3f9658adebbdf8508d325528c3206f16#diff-364713d7776956cb8b0a771e9b62f82dL1023.
So my last commit try to fix this as your guidance, keep the same behavior of local scheme file before https://github.com/apache/spark/commit/c4b1108c3f9658adebbdf8508d325528c3206f16, please help me check whether it is the correct semantics we want. Thanks!

**Comment by xuanyuanking** [comment]
@jerryshao Great thanks for your review and detailed explain, based on your guidance, I found the behavior about the file in `local` scheme added in fileServer was introduced by the PR https://github.com/apache/spark/commit/c4b1108c3f9658adebbdf8508d325528c3206f16, and before that, the file in `local` scheme is treated as local files: https://github.com/apache/spark/commit/c4b1108c3f9658adebbdf8508d325528c3206f16#diff-364713d7776956cb8b0a771e9b62f82dL1023.
So my last commit try to fix this as your guidance, keep the same behavior of local scheme file before https://github.com/apache/spark/commit/c4b1108c3f9658adebbdf8508d325528c3206f16, please help me check whether it is the correct semantics we want. Thanks!

**Comment by xuanyuanking** [comment]
retest this please.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 5]
```diff
@@ -1517,9 +1517,19 @@ class SparkContext(config: SparkConf) extends Logging {
...
     val schemeCorrectedPath = uri.getScheme match {
-      case null | "local" => new File(path).getCanonicalFile.toURI.toString
+      case null =>
+        new File(path).getCanonicalFile.toURI.toString
+      case "local" =>
+        localFile = true
+        val tmpPath = new File(uri.getPath).getCanonicalFile.toURI.toString
+        // SPARK-24195: Local is not a valid scheme for FileSystem, we should only keep path here.
+        uri = new Path(uri.getPath).toUri
+        tmpPath
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: I think the change here make file with "local" scheme a no-op. This makes me think whether supporting "local" scheme in `addFile` is meaningful or not? Because file with "local" scheme is already existed on every node, also it should be aware by the user, so adding it seems not meaingful.

By looking at the similar method `addJar`, there "local" jar is properly treated without adding to fileServer, and properly convert to the right scheme used by classloader.

Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: `This makes me think whether supporting "local" scheme in addFile is meaningful or not? Because file with "local" scheme is already existed on every node, also it should be aware by the user, so adding it seems not meaingful.`
Yeah, agree with you. The last change wants to treat "local" file without adding to fileServer and correct its scheme to "file:", but maybe add a local file, the behavior itself is a no-op? So we just forbidden user pass a file with "local" scheme in addFile?

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 6]
```diff
@@ -1519,7 +1519,12 @@ class SparkContext(config: SparkConf) extends Logging {
   def addFile(path: String, recursive: Boolean): Unit = {
     val uri = new Path(path).toUri
     val schemeCorrectedPath = uri.getScheme match {
-      case null | "local" => new File(path).getCanonicalFile.toURI.toString
+      case null => new File(path).getCanonicalFile.toURI.toString
+      case "local" =>
+        logWarning("We do not support add a local file here because file with local scheme is " +
+          "already existed on every node, there is no need to call addFile to add it again. " +
+          "(See more discussion about this in SPARK-24195.)")
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Can we please rephrase to "File with 'local' scheme is not supported to add to file server, since it is already available on every node."?

**Comment by jerryshao** [comment]
I think maybe we could:

1) either ignore the files with "local" scheme, and let user to decide how to fetch the files, like what current fix.
2) or copy the 'local' scheme files to the `SparkFiles#getRootDirectory` both in driver and executor. The change would be in `Utils#fetchFile`.

@jiangxb1987 @vanzin what's your option?

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 6 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: Got it, rephrase done.

**Comment by xuanyuanking** [comment]
retest this please.

**Review by srowen — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
retest this please.

**Comment by jiangxb1987** [comment]
Please also update the title and PR description because we changed the proposed solution in the middle.

**Comment by xuanyuanking** [comment]
@jiangxb1987 Thanks for reminding, rephrase done.

**Review by jerryshao — APPROVED** [review]
LGTM, @jiangxb1987 WDYT?

**Comment by jiangxb1987** [comment]
lgtm too

**Comment by HyukjinKwon** [comment]
SGTM too

**Comment by jerryshao** [comment]
Merging to master branch. Thanks all!

**Comment by xuanyuanking** [comment]
Thanks everyone for your help!
